src/Query.ts

```diff
@@ -0,0 +1,164 @@
+import { TaskEither, taskEither } from 'fp-ts/lib/TaskEither';
+import { Cache } from './Cache';
+import { mapWithKey, sequence, map as mapRecord } from 'fp-ts/lib/Record';
+import {
+  Strategy,
+  JSON,
+  setoidStrict,
+  setoidShallow,
+  setoidJSON,
+  refetch
+} from './Strategy';
+import { Setoid } from 'fp-ts/lib/Setoid';
+import { CacheValue, getSetoid } from './CacheValue';
+import {
+  EnforceNonEmptyRecord,
+  ObservableQueries,
+  ProductA,
+  ProductL,
+  ProductP
+} from './util';
+
+export type Fetch<A, L, P> = (input: A) => TaskEither<L, P>;
+
+interface BaseQuery<A, L, P> {
```

Comment:
if A, L and P represent input, error and return value, how do we distinguish two identical queries with different endpoints?